### ISSUE TITLE
Kubernetes unnecessarily requiring quoting on numeric env vars

### ISSUE THREAD

**Issue by wyardley** [kubernetes/kubernetes#82296]
**What happened**:
```
Error from server (BadRequest): error when creating "/tmp/test.yaml": Deployment in version "v1" cannot be handled as a Deployment: json: cannot unmarshal number into Go struct field EnvVar.spec.template.spec.containers.env.value of type string
```

**What you expected to happen**:
For the deployment to get applied

**How to reproduce it (as minimally and precisely as possible)**:
```yaml
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: foo
spec:
  selector:
    matchLabels:
      app: foo
  template:
    metadata:
      labels:
        app: foo
    spec:
      containers:
        - name: nginx
          image: nginx:latest
          env:
            - name: PROJECT_VERSION
              value: 03788745
```

**Anything else we need to know?**:
libraries like jsyaml and PyYAML etc. write out YAML 1.1 by only quoting numeric strings when necessary. So, a number with a leading 0 _and_ which contains non-octal characters (e.g., 8 or 9) will not be surrounded with quotes when written out (this is an issue when using these tools to template configs).

https://github.com/nodeca/js-yaml/issues/394

It seems like `kubectl` accepts these values, however, the Kubernetes server requires them to be quoted, even when, according to my reading of things, the string does not need to be quoted. I'm not sure if this is a bug in one of the Go YAML parsing libraries or elsewhere. Based on below, doesn't seem as if YAML 1.2 is supported yet (see issue link below), though I think the rules may be slightly different with 1.2.

Maybe semi-related: https://github.com/kubernetes/kubernetes/issues/34146
https://github.com/helm/helm/issues/2693

**Environment**:
- Kubernetes version (use `kubectl version`):
```
Client Version: v1.24.0
Kustomize Version: v4.5.4
Server Version: v1.23.5-gke.1503
```
- Cloud provider or hardware configuration: Google (GKE)
- OS (e.g: `cat /etc/os-release`): OS X (client), Linux (server)
- Kernel (e.g. `uname -a`): n/a; let me know if you need more details
- Install tools:
- Network plugin and version (if this is a network-related bug):
- Others:

**Comment by liggitt** [comment]
This is an issue in our yaml conversion library.

It also looks like the library is treating numbers with leading 0's as decimal numbers (the error message says `"value":3788745`, which is even more concerning.

We need to determine the current behavior of the library's handling of non-decimal number formats. For example:
* users specifying numeric fields as values with leading 0's (e.g. `defaultMode` fields in PodSpec)
* users specifying string fields as unquoted digit-only values with leading 0's (this bug)

Once we understand the current behavior, and what values are currently accepted, we can determine what we can safely modify (and over what time-frame), considering backwards-compatibility with current users, and cross-compatibility with other yaml-producing libraries

**Comment by kikisdeliveryservice** [comment]
hi @liggitt
 
Bug Triage Release Team checking in. Code freeze for the 1.17 release is on November 18. Is this still intended for 1.17? Thanks!

**Comment by markyjackson-taulia** [comment]
Hi @liggitt

Bug Triage Release Team checking in. Will this be staffed for v1.18

**Comment by maikelvl** [comment]
We've encountered this issue today in a similar way.

```yaml
env:
- name: MY_ENV
  value: 12345678e2
```

Works on v1.15.3, but fails on v1.15.9:

```
Error from server (BadRequest): error when creating "test.yaml": Job in 
version "v1" cannot be handled as a Job: v1.Job.Spec: v1.JobSpec.Template: 
v1.PodTemplateSpec.Spec: v1.PodSpec.Containers: []v1.Container: 
v1.Container.Env: []v1.EnvVar: v1.EnvVar.Value: ReadString: expects " or n, but 
found 7, error found in #10 byte of ...|,"value": 1234567800}|..., bigger 
context ...|":"MY_ENV","value": 1234567800}...
```


A fix is to quote the value:
```yaml
env:
- name: MY_ENV
  value: "12345678e2"
```

**Comment by markyjackson-taulia** [comment]
Hi @liggitt

Gentle ping to see if this is staffed for v1.18

**Comment by liggitt** [comment]
@maikelvl your issue is distinct. 12345678e2 is a valid number (it's numeric value is 1234567800), so you need to quote it to specify it in a string field.

@markyjackson-taulia it is not yet staffed. higher priority issues are being worked on currently.

**Comment by markyjackson-taulia** [comment]
@liggitt are we ok to move this into v1.19 milestone or remove the milestone until staffing is available?

**Comment by markyjackson-taulia** [comment]
@liggitt ok to move this to v1.19
![image](https://user-images.githubusercontent.com/16655670/74669028-36b61d80-515b-11ea-8c66-9d00c0905766.png)

**Comment by liggitt** [comment]
yes, upstream work on the yaml libraries is not planned for 1.18

**Comment by maikelvl** [comment]
> @maikelvl your issue is distinct. 12345678e2 is a valid number (it's numeric value is 1234567800), so you need to quote it to specify it in a string field.

I agree. My point is: this behaviour has changed in patch releases:

**kubectl v1.15.4**
```console
$ kubectl get configmap test -o yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: test
  ...
data:
  test: 12345678e2
```

---

**kubectl v1.15.5**
```console
$ kubectl get configmap test -o yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: test
  ...
data:
  test: "12345678e2"
```

**Comment by maikelvl** [comment]
@liggitt Do you agree such change in a patch release should not happen? ⬆️

**Comment by maikelvl** [comment]
As a workaround I use `sed -E 's|(\w+):\s+(\d+e\d+)$|\1: "\2"|g'`

<details>

<pre class="highlight highlight-source-shell">
$ cat << EOF > configmap.yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: test
data:
  test: 123456e78
EOF
</pre>

### Problem

<pre>
$ # kubectl v1.15.5 
$ cat configmap.yaml | kubectl apply -f -

Error from server: error when applying patch:
{"data":{"test":1.23456e+83},"metadata":{"annotations":{"kubectl.kubernetes.io/last-applied-configuration":"{\"apiVersion\":\"v1\",\"data\":{\"test\":1.23456e+83},\"kind\":\"ConfigMap\",\"metadata\":{\"annotations\":{},\"name\":\"test\",\"namespace\":\"operations\"}}\n"}}}
to:
Resource: "/v1, Resource=configmaps", GroupVersionKind: "/v1, Kind=ConfigMap"
Name: "test", Namespace: "operations"
Object: &{map["apiVersion":"v1" "data":map["test":"12345678e2"] "kind":"ConfigMap" "metadata":map["annotations":map["kubectl.kubernetes.io/last-applied-configuration":"{\"apiVersion\":\"v1\",\"data\":{\"test\":\"12345678e2\"},\"kind\":\"ConfigMap\",\"metadata\":{\"annotations\":{},\"name\":\"test\",\"namespace\":\"operations\"}}\n"] "creationTimestamp":"2020-03-23T11:18:02Z" "name":"test" "namespace":"operations" "resourceVersion":"98824733" "selfLink":"/api/v1/namespaces/operations/configmaps/test" "uid":"abcff6f5-14ed-4fb3-90a4-9b6b30397d04"]]}
for: "STDIN": unrecognized type: string
</pre>

### Workaround

<pre>
$ # kubectl v1.15.5 
$ cat configmap.yaml | sed -E 's|(\w+):\s+(\d+e\d+)$|\1: "\2"|g' | kubectl apply -f -
configmap/test configured
</pre>

</details>

**Comment by liggitt** [comment]
@maikelvl that was a bug fix in 1.15.5, due to upgrading the yaml library kubernetes uses. That is allowed in a patch release. What was output in 1.15.4 would not work properly with kubectl create or replace. What was output in 1.15.5 would work properly.

**Comment by maikelvl** [comment]
@liggitt Thank you for your clarification. 

---

For reference: the mentioned yaml library upgrade is applied in https://github.com/kubernetes/kubernetes/pull/83261/files

**Comment by mkorbi** [comment]
Heyho, I see that there are a couple of merges was done, is this still planned for 1.20?
@liggitt

**Comment by liggitt** [comment]
No, there are no plans or work in progress on this issue as far as I am aware.

**Comment by mkorbi** [comment]
I think than we can move it to 1.21?

**Comment by sayanchowdhury** [comment]
Hi 👋🏽 , Bug Triage here, We're pretty late into code freeze, and there is not much activity in this issue so removing the 1.20 milestone.

/milestone clear

**Comment by wyardley** [comment]
@liggitt @sayanchowdhury this has now not made it into 1.17, 1.18, 1.19, or 1.20, and has been open for over a year... Any idea when someone will be able to work on this?

**Comment by neolit123** [comment]
if there are contributors that wish to fix this ASAP, PRs for the library should go at http://sig.k8s.io/yaml.
for review / approval of proposals / PRs, the owners of the library are sig-api-machinery:
https://github.com/kubernetes/community/tree/master/sig-api-machinery

**Comment by liggitt** [comment]
> Any idea when someone will be able to work on this?

Not really. This remains in the middle of the backlog, priority-wise.

The issue lives in the yaml libraries we use, but as noted in https://github.com/kubernetes/kubernetes/issues/82296#issuecomment-529675294, changing how we interpret existing data has compatibility implications that are far-reaching.

> We need to determine the current behavior of the library's handling of non-decimal number formats. For example:
>
> * users specifying numeric fields as values with leading 0's (e.g. defaultMode fields in PodSpec)
> * users specifying string fields as unquoted digit-only values with leading 0's (this bug)
>
> Once we understand the current behavior, and what values are currently accepted, we can determine what we can safely modify (and over what time-frame), considering backwards-compatibility with current users, and cross-compatibility with other yaml-producing libraries

I would expect any work done in the yaml libraries to address this to pay similar attention to compatibility implications for data which is currently accepted/parsed.

**Comment by budnik** [comment]
Hi. 

I wonder If issue [mentioned](https://github.com/kubernetes/kubernetes/issues/82296#issuecomment-583121688) by @maikelvl has to be reported as separate issue? According to original report it seem like a regression in 1.15.x

**Comment by yakovm44** [comment]
Hi
I found solution that work with clean numbers

env:
  - name: "abcd"
    value: "\"111\""

**Comment by withlin** [comment]
same issue.

**Comment by brianpursley** [comment]
I spent some time looking at this today and I think I see what is causing this problem.  I'm not quite sure the best way to go about fixing it though.  The trick is fixing the problem for YAML without affecting JSON because they share much of the same code.

I'll leave what I found here in case it helps.

---

### The Cause

The problem seems to occur in `staging/src/k8s.io/apimachinery/pkg/runtime/serializer/json/json.go` in [`Decode()`](https://github.com/kubernetes/kubernetes/blob/122d6352452f3f4a2d52bdb8b114b6e0ec49358e/staging/src/k8s.io/apimachinery/pkg/runtime/serializer/json/json.go#L201-L294)

When you pass YAML, it is [first converted to JSON as an intermediate step](https://github.com/kubernetes/kubernetes/blob/a68644ba24b84b74089f5d0b70576725c38d7060/staging/src/k8s.io/apimachinery/pkg/runtime/serializer/json/json.go#L212) and then later is unmarshalled to the target type.

However, the YAML-to-JSON intermediate step conversion is unaware of which type each field will eventually need to be decoded into and the call to `yaml.YAMLToJSON()` emits unquoted JSON numeric values for numeric values in the YAML.

### Example

To illustrate what I'm talking about, consider this YAML fragment passed into `Decode()`:
```
env:
- name: foo
- value: 1
```

Because it is YAML, an intermediate step converts the YAML to JSON, resulting in this:
```
"env": {
  "name": "foo",
  "value": 1
}
```

Now consider that it is going to unmarshal into this type below (which is somewhat similar to the real EnvVar struct, just cut down).  Notice `Value` is a string, which is what triggers the problem to occur:
```
type EnvVar struct {
    name string `json:"name" protobuf:"bytes,1,opt,name=name"`
    Value string `json:"value,omitempty" protobuf:"bytes,2,opt,name=value"`
}
```

When `Decode()` calls `Unmarshal()`, it fails with the following message because ReadString is attempted on the number `1` and not the string `"1"`.

*Losing the leading zero is just another symptom of this same problem.  When the YAML-to-JSON conversion happens, it becomes an unquoted number in JSON and therefore loses the leading zero.*

### Unit tests

I created a couple of unit test cases for [`TestDecode`](https://github.com/kubernetes/kubernetes/blob/a68644ba24b84b74089f5d0b70576725c38d7060/staging/src/k8s.io/apimachinery/pkg/runtime/serializer/json/json_test.go#L384) which show the issue.  These tests will fail if you plug them in.  They are not committed, but I'm including them here as extra info in case it helps someone troubleshoot the problem.
```
 		// Yaml integer can be decoded into a string
		{
			data:        []byte("kind: Test\napiVersion: other/blah\nOther: 1"),
			into:        &testDecodable{},
			typer:       &mockTyper{err: runtime.NewNotRegisteredErrForKind("mock", schema.GroupVersionKind{Kind: "Test", Group: "other", Version: "blah"})},
			expectedGVK: &schema.GroupVersionKind{Kind: "Test", Group: "other", Version: "blah"},
			expectedObject: &testDecodable{
				Other: "1",
			},
			yaml: true,
		},

		// Yaml float can be decoded into a string
		{
			data:        []byte("kind: Test\napiVersion: other/blah\nOther: 1.234"),
			into:        &testDecodable{},
			typer:       &mockTyper{err: runtime.NewNotRegisteredErrForKind("mock", schema.GroupVersionKind{Kind: "Test", Group: "other", Version: "blah"})},
			expectedGVK: &schema.GroupVersionKind{Kind: "Test", Group: "other", Version: "blah"},
			expectedObject: &testDecodable{
				Other: "1.234",
			},
			yaml: true,
		},
```

---

I hope this helps whoever else is looking into this issue.

I'd be glad to work on a fix for this, but as noted above it is going to be a little tricky to figure out how to fix this for YAML inputs without affecting JSON inputs, so I would want to get some direction on how best to do that.  I am assuming that in cases where JSON is passed in, it would still be desirable to keep the current behavior of failing if you try to pass a number in for a string.

**Comment by wyardley** [comment]
Hi all

It's been many moons, so, in hopes that someone, somewhere is still listening, I've updated the description with an `apps/v1` style repro case and the error / version output from a newer version of Kubernetes.

**Comment by wyardley** [comment]
> any workaround???? I need to declare numbers leading with 0

I think the workaround is to quote the value, unless there's some reason that doesn't work for your use case?

**Comment by thockin** [comment]
https://twitter.com/thockin/status/1346873150382870528

TLDR:

We could coerce numbers -> strings when we are unmarshalling, I guess?  I don't know if that's standard.

Octal in YAML is broken, do not rely on it (sad!)

Should we recognize 0 followed by digits where one is > 7 as a string?  That really feels like a stretch to me.

**Comment by kchrideq** [comment]
With "|"  that's work

`metadata:
 namespace: metallb-system
 name: config
data:
 config: |
  address-pools:
  - name:	default
    protocol:	layer2
    addresses:
    - 192.168.59.95-192.168.59.125`

**Comment by kchrideq** [comment]
With "|" that works.
```yaml
metadata:
 namespace: metallb-system
 name: config
data:
 config: |
  address-pools:
  - name:	default
    protocol:	layer2
    addresses:
    - 192.168.59.95-192.168.59.125
    ```

**Comment by Hazem-Al-Aqailah** [comment]
this problem still persists even when i apply double quotes to the numerical values
i get this error logs when trying to apply my deployment:
`Error from server (BadRequest): error when creating "deployment.yml": Deployment in version "v1" cannot be handled as a Deployment: json: cannot unmarshal bool into Go struct field EnvVar.spec.template.spec.containers.env.value of type string`
my yml file is as follows:

`apiVersion: apps/v1
kind: Deployment
metadata:
  labels:
    app: keycloak
  name: keycloak
  namespace: kc
spec:
  replicas: 1
  selector:
    matchLabels:
      app: keycloak
  template:
    metadata:
      labels:
        app: keycloak
    spec:
      containers:
        - name: keycloak-container
          image: quay.io/keycloak/keycloak:19.0.1
          command:
            - "start-dev --import-realm"
          imagePullPolicy: IfNotPresent
          env:
            - name: KEYCLOAK_ADMIN
              value: admin
            - name: KEYCLOAK_ADMIN_PASSWORD
              value: admin
            - name: KC_FEATURES
              value: scripts
            - name: KC_HTTP_PORT
              value: "9080"
            - name: KC_HTTPS_PORT
              value: "9443"
            - name: KC_HTTPS_ENABLED
              value: true
            - name: KC_HTTP_ENABLED
              value: true
            - name: KC_DB
              value: mariadb
            - name: KC_DB_USERNAME
              value: keycloak
            - name: KC_DB_PASSWORD
              value: keycloak
            - name: KC_DB_SCHEMA
              value: keycloak
            - name: KC_DB_URL
              value: jdbc:mysql://<vm-ip>:3306/keycloak?useSSL=false
          ports:
            - containerPort: 9080
          volumeMounts:
            - mountPath: /opt/jboss/keycloak/realmconfig
              name: keycloak-vol
      terminationGracePeriodSeconds: 0
      volumes:
        - name: keycloak-vol
          persistentVolumeClaim:
            claimName: keycloackclaim`
the problem should be fixed when applying double quotes to the following env variables:
KC_HTTPS_PORT
KC_HTTP_PORT
but still facing the same error
ps: I'm using minikube v1.27.1 on a centos 7 vm
any help would be appreciated

**Comment by kchrideq** [comment]
`cannot unmarshal bool into Go struct field EnvVar.spec.template.spec.containers.env.value of type string`

It says you must have a string, while the value is a bool (true/false). So the issu is probabley in KC_HTTPS_ENABLED or KC_HTTP_ENABLED
Try applying double quotes to true values

**Comment by thockin** [comment]
Sadly, the conversion process converts YAML `true` to JSON `true` and unmarshalling `true` into a string is not a thing we do (though I can comprehend an argument for "maybe we should")

**Comment by bravikr** [comment]
Is this issue fixed ?

I still the issue with the below yaml file and  version

Deployment in version “v1” cannot be handled as a Deployment: strict decoding error: unknown field “metadata.spec”

Get the above error with latest version of kubernetes

Engine:
Version: 24.0.2
API version: 1.43 (minimum version 1.12)
Go version: go1.20.4
Git commit: 659604f
Built: Thu May 25 21:52:17 2023
OS/Arch: linux/amd64
Experimental: false
containerd:
Version: 1.6.21
GitCommit: 3dce8eb055cbb6872793272b4f20ed16117344f8
runc:
Version: 1.1.7
GitCommit: v1.1.7-0-g860f061
docker-init:
Version: 0.19.0
GitCommit: de40ad0

Client Version: v1.25.9
Kustomize Version: v4.5.7
Server Version: v1.26.3

trying to do the deployment using below yaml file

apiVersion: apps/v1
kind: Deployment
metadata:
name: mongodb-deployment
labels:
app: mongodb
spec:
replicas: “1”
selector:
matchLabels:
app: mongodb
template:
metadata:
labels:
app: mongodb
spec:
containers:

name: mongodb
image: mongo
ports:
containerPort: “27107”
env:
name: MONGO_INITDB_ROOT_USERNAME
valueFrom:
secretKeyRef:
name: mongodb-secret
key: mongo-root-username
name: MONGO_INITDB_ROOT_PASSWORD
valueFrom:
secretKeyRef:
name: mongodb-secret
key: mongo-root-password

**Comment by thockin** [comment]
@bravikr - the YAML you pasted has lost all indenting, and whitespace matters.

Given the error (metadata.spec) I am 100% confident you have an indenting problem.

**Comment by bravikr** [comment]
I did tried with many many ways to do and check the same.
since few days correcting and i feel I am done . :)
[mongo.zip](https://github.com/kubernetes/kubernetes/files/11877762/mongo.zip)
I tried by going through line by line..and still couldnt do that..

**Comment by thockin** [comment]
Put three backticks on a line, the your YAML.

I will edit your post so you can see

**Comment by bravikr** [comment]
@thockin
 could you try to post the yaml file, I tried so many times about editing the file and couldnt make it.

**Comment by thockin** [comment]
Here's the YAML in your zipfile, and indeed thindenting is wrong:

```
apiVersion: apps/v1
kind: Deployment
metadata:
        name: mongodb-deployment
        labels:
          app: mongodb
        spec:
          replicas: "1"
          selector:
            matchLabels:
              app: mongodb
          template:
            metadata:
              labels:
                app: mongodb
            spec:
              containers:
              - name: mongodb
                image: mongo
                ports:
                - containerPort: "27107"
                env:
                - name: MONGO_INITDB_ROOT_USERNAME
                  valueFrom:
                    secretKeyRef:
                      name: mongodb-secret
                      key: mongo-root-username
                - name: MONGO_INITDB_ROOT_PASSWORD
                  valueFrom:
                    secretKeyRef:
                      name: mongodb-secret
                      key: mongo-root-password
```

`spec` and everything below it are inside metadata, which is incorrect.

**Comment by thockin** [comment]
Try converting it to JSON, and you'll see what I mean - the braces make it more obvious.

**Comment by bravikr** [comment]
[new_yaml_mongodb.zip](https://github.com/kubernetes/kubernetes/files/11924186/new_yaml_mongodb.zip)

yes. you are right.
here is the working one.

converting to json did the trick and visibility

**Comment by thockin** [comment]
To this original bug:

1) Octal is YAML is broken.  Don't use it.   https://twitter.com/thockin/status/1346873150382870528

2) While we could (in theory) sniff incoming JSON at the API server, and parse things like numbers and booleans into strings when the schema demands a string, a) nobody I know of does that; b) we don't always do schemaful parsing, so that logic would have to be replicated any time we parse into a schema'd type; c) we also support protobuf as an encoding, and I think that is nigh impossible to fix the same way.

3) We could MAYBE do this coercion only when converting from YAML to JSON (client-side) but then it would be surprising and inconsistent.

So I think the lack of movement and the lack of possible solutions means, this is "working as intended", even if we wish it were nicer.
